packages/core/src/client.ts

```diff
@@ -1969,6 +1970,31 @@ export class MedplumClient extends EventTarget {
       this.createResource(resource, options)) as Promise<T>;
   }
 
+  /**
+   * Upsert a resource: update it in place if it exists, otherwise create it.  This is done in a single, transactional
+   * request to guarantee data consistency.
+   * @param resource - The resource to update or create.
+   * @param query - A FHIR search query to uniquely identify the resource if it already exists.
+   * @param options  - Optional fetch options.
+   * @returns The updated/created resource.
+   */
+  async upsert<T extends Resource>(resource: T, query: string, options?: MedplumRequestOptions): Promise<T> {
```

Comment:
@rahul1  - thoughts on function name?

At minimum, I think we should use `upsertResource()` to follow the pattern of `createResource()`, `updateResource()`, etc, and to give us room in the future for other potential upsert methods (??).

I'd also be ok to reuse the existing `createResourceIfNoneExist()` for smoother migration.  Not sure if the API contract is similar enough to reuse though.